homeassistant/components/dreo/coordinator.py

```diff
@@ -0,0 +1,144 @@
+"""Data update coordinator for Dreo devices."""
+
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from typing import Any, NotRequired, TypedDict
+
+from hscloud.const import DEVICE_TYPE, FAN_DEVICE
+from hscloud.hscloud import HsCloud
+from hscloud.hscloudexception import HsCloudException
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util.percentage import ranged_value_to_percentage
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=10)
+
+
+class DreoGenericDeviceData(TypedDict):
+    """Base data for all Dreo devices."""
+
+    available: bool
+    is_on: bool
+
+
+class DreoFanDeviceData(DreoGenericDeviceData):
+    """Data specific to Dreo fan devices."""
+
+    mode: NotRequired[str]
+    oscillate: NotRequired[bool]
+    speed_percentage: NotRequired[int]
+
+
+DreoDeviceData = DreoFanDeviceData | DreoGenericDeviceData
+
+
+class DeviceDataFactory:
+    """Factory to create device data based on device type."""
+
+    @staticmethod
+    def create_data(
+        coordinator: DreoDataUpdateCoordinator,
+        base_data: DreoGenericDeviceData,
+        status: dict[str, Any],
+    ) -> DreoDeviceData:
+        """Create appropriate device data based on device type."""
+        if (
+            coordinator.device_type == FAN_DEVICE.get("type")
+            and coordinator.device_model
+        ):
+            return FanDataStrategy.process_data(
+                coordinator.device_model, base_data, status
+            )
+        return DreoFanDeviceData(**base_data)
+
+
+class FanDataStrategy:
+    """Strategy for processing fan device data."""
+
+    @staticmethod
+    def process_data(
```

Comment:
what about moving this method to `DreoFanDeviceData` and removing `FanDataStrategy`, so that we don't split the data processing from the data storage classes?